Ag develop

a-petelin opened the pull request.

Commit:
Add debian package generating

a-petelin committed:
Add debian package generating

a-petelin committed:
Fix missed include for cpack

Commit:
Fix missed include for cpack

Commit:
Fix incorrect message on checking cast between NsSkelJson types
 On branch 1-feature-add-use-config-from-zookeeper
	modified:   ns-skeleton/src/NsSkelJson.cpp

a-petelin committed:
~ Add use zookeeper as config source
	type problems now
 On branch 1-feature-add-use-config-from-zookeeper
	modified:   ns-skeleton/CMakeLists.txt
	new file:   ns-skeleton/include/AbstractConfig.h
	new file:   ns-skeleton/include/CVInvalidType.h
	new file:   ns-skeleton/include/CVNoNodeData.h
	new file:   ns-skeleton/include/CVTooLongData.h
	new file:   ns-skeleton/include/CVTooManyChildren.h
	new file:   ns-skeleton/include/ConfigValidator.h
	modified:   ns-skeleton/include/NsSkelConfiguration.h
	modified:   ns-skeleton/include/NsSkelRpcService.h
	new file:   ns-skeleton/include/file/FileConfig.h
	new file:   ns-skeleton/include/zoo/ZooConfig.h
	new file:   ns-skeleton/src/CVInvalidType.cpp
	new file:   ns-skeleton/src/CVNoNodeData.cpp
	new file:   ns-skeleton/src/CVTooLongData.cpp
	new file:   ns-skeleton/src/CVTooManyChildren.cpp
	new file:   ns-skeleton/src/ConfigValidator.cpp
	modified:   ns-skeleton/src/NsSkelConfiguration.cpp
	modified:   ns-skeleton/src/NsSkelRpcRegistry.cpp
	modified:   ns-skeleton/src/NsSkelRpcService.cpp
	new file:   ns-skeleton/src/file/FileConfig.cpp
	new file:   ns-skeleton/src/zoo/ZooConfig.cpp

a-petelin committed:
~ Fix test
 On branch 1-feature-add-use-config-from-zookeeper
	modified:   ns-tests/testExceptionThrowThrough.cpp

a-petelin committed:
~ Fix build with config from file source
 On branch 1-feature-add-use-config-from-zookeeper
	modified:   ns-skeleton/include/NsCmdLineParameters.h
	modified:   ns-skeleton/include/file/FileConfig.h
	modified:   ns-skeleton/src/NsCmdLineParameters.cpp
	modified:   ns-skeleton/src/file/FileConfig.cpp

a-petelin committed:
Add use version for deb-package from VERSION.txt
 On branch 1-feature-add-use-config-from-zookeeper
	modified:   VERSION.txt
	modified:   ns-skeleton/CMakeLists.txt

a-petelin committed:
Fix type problem in zookeeper source of config
 On branch 1-feature-add-use-config-from-zookeeper
	modified:   CMakeLists.txt
	modified:   ns-skeleton/CMakeLists.txt
	modified:   ns-skeleton/include/ConfigValidator.h
	modified:   ns-skeleton/include/zoo/ZooConfig.h
	modified:   ns-skeleton/src/ConfigValidator.cpp
	modified:   ns-skeleton/src/file/FileConfig.cpp
	modified:   ns-skeleton/src/zoo/ZooConfig.cpp

a-petelin committed:
Refactor work with command line parameters
 On branch 1-feature-add-use-config-from-zookeeper
	modified:   ns-skeleton/include/NsCmdLineParameters.h
	modified:   ns-skeleton/src/NsCmdLineParameters.cpp
	modified:   ns-skeleton/src/NsSkelConfiguration.cpp
	modified:   ns-skeleton/src/NsSkelRpcServer.cpp
	modified:   ns-skeleton/src/NsSkeleton.cpp

a-petelin committed:
Add misc project
 On branch 1-feature-add-use-config-from-zookeeper
	new file:   misc/CMakeLists.txt
	new file:   misc/main.cpp

a-petelin committed:
Refactor version of project
 On branch 1-feature-add-use-config-from-zookeeper
	modified:   VERSION.txt

a-petelin committed:
Refactor generating debian packages for framework
 On branch 1-feature-add-use-config-from-zookeeper
	modified:   VERSION.txt
	modified:   ns-skeleton/CMakeLists.txt

a-petelin committed:
Add misc tool ns-configurator
 On branch 1-feature-add-use-config-from-zookeeper
	modified:   CMakeLists.txt
	modified:   misc/CMakeLists.txt
	new file:   misc/ns-configurator/CMakeLists.txt
	new file:   misc/ns-configurator/VERSION.txt
	renamed:    misc/main.cpp -> misc/ns-configurator/main.cpp
	modified:   ns-skeleton/CMakeLists.txt

a-petelin committed:
Refactor README.md add more info

a-petelin committed:
Fix use relative path to other components directories
 On branch 1-feature-add-use-config-from-zookeeper
	modified:   misc/ns-configurator/CMakeLists.txt
	modified:   ns-tests/CMakeLists.txt

a-petelin committed:
Fix missing project definition in ns-configurator
 On branch 1-feature-add-use-config-from-zookeeper
	modified:   CMakeLists.txt
	modified:   misc/ns-configurator/CMakeLists.txt

a-petelin committed:
Fix misspelled description in misc component package definition

a-petelin committed:
Merge branch 'ag-develop' of https://github.com/dmerkushov/nanoservices into ag-develop

a-petelin committed:
Fix incorrect mail for maintainer

a-petelin committed:
Add setUp and tearDown for ns-tests by use ns-configurator
 On branch ag-develop
	modified:   ns-tests/CMakeLists.txt
	new file:   ns-tests/json-file.txt
	new file:   ns-tests/json-zoo.txt

a-petelin committed:
Fix corrupt input json by space splitting
 On branch ag-develop
	modified:   misc/ns-configurator/main.cpp

a-petelin committed:
Fix subproject order
 On branch ag-develop
	modified:   CMakeLists.txt

a-petelin committed:
Fix crash on config with descriptions
 On branch ag-develop
	modified:   ns-skeleton/include/AbstractConfig.h
	modified:   ns-skeleton/include/file/FileConfig.h
	modified:   ns-skeleton/include/zoo/ZooConfig.h
	modified:   ns-skeleton/src/NsSkelConfiguration.cpp
	modified:   ns-skeleton/src/file/FileConfig.cpp
	modified:   ns-skeleton/src/zoo/ZooConfig.cpp

a-petelin committed:
Add tests for config
 On branch ag-develop
	modified:   ns-skeleton/CMakeLists.txt
	new file:   ns-skeleton/test/AbstractConfigTest.h
	new file:   ns-skeleton/test/file/TestActionCheck.h
	new file:   ns-skeleton/test/zoo/TestActionCheck.h

a-petelin committed:
Fix commands for launch ns-configurator
 On branch ag-develop
	modified:   ns-tests/CMakeLists.txt

a-petelin committed:
Refactor zookeeper specified part of test
	define PREFIX (will be used definition from cmake leater)
 On branch ag-develop
	modified:   ns-skeleton/test/zoo/TestActionCheck.h

a-petelin committed:
~ Todo for help message in ns-configurator
 On branch ag-develop
	modified:   misc/ns-configurator/main.cpp

a-petelin committed:
Refactor class for config from file source
	Full CRUD now, behavior same as class for config from zookeeper
 On branch ag-develop
	modified:   ns-skeleton/include/file/FileConfig.h
	modified:   ns-skeleton/src/file/FileConfig.cpp
	modified:   ns-skeleton/test/AbstractConfigTest.h
	modified:   ns-skeleton/test/file/TestActionCheck.h

a-petelin committed:
Refactor test config for ns-tests
 On branch ag-develop
	modified:   ns-tests/CMakeLists.txt
	deleted:    ns-tests/json-file.txt
	deleted:    ns-tests/json-zoo.txt
	new file:   ns-tests/json.txt

a-petelin committed:
Refactor move string_split to new file from NsCmdLineParameters.cpp
 On branch ag-develop
	modified:   misc/ns-configurator/CMakeLists.txt
	modified:   ns-skeleton/CMakeLists.txt
	modified:   ns-skeleton/include/NsCmdLineParameters.h
	new file:   ns-skeleton/include/NsUtils.h
	modified:   ns-skeleton/src/NsCmdLineParameters.cpp
	new file:   ns-skeleton/src/NsUtils.cpp
	modified:   ns-skeleton/src/file/FileConfig.cpp
	modified:   ns-skeleton/src/zoo/ZooConfig.cpp

a-petelin committed:
Merge branch 'master' of https://github.com/dmerkushov/nanoservices into ag-develop

a-petelin committed:
Refactor command line parameter parse from getopt to argp
 On branch ag-develop
	modified:   CMakeLists.txt
	modified:   misc/ns-configurator/CMakeLists.txt
	modified:   misc/ns-configurator/main.cpp
	modified:   ns-skeleton/include/NsCmdLineParameters.h
	modified:   ns-skeleton/src/NsCmdLineParameters.cpp
	modified:   ns-skeleton/src/file/FileConfig.cpp
	modified:   ns-tests/json.txt

a-petelin committed:
Add ns-shutdown and blank testService to misc
 On branch ag-develop
	modified:   misc/CMakeLists.txt
	new file:   misc/ns-shutdown/CMakeLists.txt
	new file:   misc/ns-shutdown/main.cpp
	new file:   misc/testService/CMakeLists.txt
	new file:   misc/testService/main.cpp

a-petelin committed:
Fix coredump then use NsSkeleton::init without command line args
 On branch ag-develop
	modified:   ns-skeleton/src/NsCmdLineParameters.cpp

a-petelin committed:
Refactor up version
 On branch ag-develop
	modified:   VERSION.txt

a-petelin committed:
Refactor testService free project now
	Add .cmake file for find_package work
 On branch ag-develop
	modified:   CMakeLists.txt
	modified:   VERSION.txt
	modified:   misc/CMakeLists.txt
	modified:   misc/testService/CMakeLists.txt
	new file:   misc/testService/README.md
	modified:   ns-skeleton/CMakeLists.txt
	new file:   ns-skeleton/FindNanoservices.cmake
	new file:   ns-skeleton/ns-skeleton.conf

a-petelin committed:
Refactor move FindNanoservices.cmake to dev component
 On branch ag-develop
	modified:   ns-skeleton/CMakeLists.txt

a-petelin committed:
Fix promise_error exception while init ZooConfig
 On branch ag-develop
	modified:   ns-skeleton/src/zoo/ZooConfig.cpp

a-petelin committed:
Add HTTP server properties into test config
 On branch ag-develop
	modified:   ns-tests/json.txt

a-petelin committed:
Add msgpack-c and httplib as third party
 On branch ag-develop
	new file:   third-party/README.md
	new folder: third-party/cpp-httplib
	new folder: third-party/msgpack-c

a-petelin committed:
Add test method into testService
 On branch ag-develop
	modified:   misc/testService/main.cpp
	new file:   misc/testService/testMethod.cpp
	new file:   misc/testService/testMethod.h

a-petelin committed:
Add using msgpack from third-party directory
 On branch ag-develop
	modified:   misc/ns-configurator/CMakeLists.txt
	modified:   misc/ns-shutdown/CMakeLists.txt
	modified:   ns-skeleton/CMakeLists.txt
	modified:   ns-tests/CMakeLists.txt

a-petelin committed:
Refactor NsSkelRpcHttpServer to use httplib
 On branch ag-develop
	modified:   ns-skeleton/include/NsSkelRpcHttpServer.h
	modified:   ns-skeleton/include/NsSkelRpcServer.h
	modified:   ns-skeleton/src/NsSkelRpcHttpServer.cpp
	modified:   ns-skeleton/src/NsSkelRpcRegistry.cpp

a-petelin committed:
Up version
 On branch ag-develop
	modified:   VERSION.txt

dmerkushov commented on CMakeLists.txt:
```diff
@@ -1,9 +1,49 @@
+elseif(ZOO)
+	SET(CPACK_PACKAGE_NAME ${CMAKE_PROJECT_NAME}-zoo)
+else()
+	message(FATAL_ERROR "You cmake it wrong!!")
```
```suggestion
	message(FATAL_ERROR "Please set FILE or ZOO config")
```

a-petelin replied:
Поменял сообщение на более понятное

dmerkushov commented on README.md:
```diff
@@ -10,18 +10,38 @@ It is in development by now, a more meaningful readme will be added later
 ------------
 
-You'll need: `cmake` (minimum version is 3.0), `make` and a C++11 compiler (the framework is tested with gcc 4.7)
+You'll need: `cmake` (minimum version is 3.0), `make`, `libzookeeper-mt2`, local instance of `zookeeper` and a C++11 compiler (the framework is tested with gcc 4.7)
```
Где взять `libzookeeper-mt2` ? Это название дебиан пакета, а не программного продукта. Он необходим для сборки? А если сборка с FILE?

a-petelin replied:
`libzookeeper-mt2` это многопоточный клиент для `zookeeper`, в принципе собирается из исходников которые идут с ним.
Если сборка идёт с FILE, то зависимости относящиеся к `zookeeper` естественно не требуются.

dmerkushov commented on README.md:
```diff
@@ -10,18 +10,38 @@ It is in development by now, a more meaningful readme will be added later
 
 Header files to include to use the library will be in `nanoservices/ns-skeleton/include/`
+
+Also successful build generate misc tool `ns-configurator` at `nanoservices/build/misc/ns-configurator`
```
```suggestion
A successful build will also generate the `ns-configurator` executable at `nanoservices/build/misc/ns-configurator`
```

dmerkushov commented on README.md:
```diff
@@ -10,18 +10,38 @@ It is in development by now, a more meaningful readme will be added later
+Build debian packages
+---------------------
+
+For build debian packages from builded sources
```
```suggestion
To build debian packages from built sources
```

dmerkushov commented on misc/ns-configurator/main.cpp:
```diff
@@ -0,0 +1,45 @@
+
+int main(int argc, char** argv) {
+	map<char, NsCmdLineParameters::opt> prms = {
+		{'p' , {"path", 'p', "PATH", 0, "Path to config node", 2}},
+		{'a' , {"data", 'a', "DATA", 0, "Data for create or update operation", 2}},
+		{'c' , {"create", 'c', 0, 0, "create config node at PATH with DATA", 1}},
+		{'r' , {"read", 'r', 0, 0, "read config node data from PATH", 1}},
+		{'u' , {"update", 'u', 0, 0, "update config node at PATH as DATA", 1}},
+		{'d' , {"delete", 'd', 0, 0, "delete config node at PATH", 1}},
```
Лучше оставить только реально поддерживаемые на данный момент, остальные закомментировать
```suggestion
//TODO		{'p' , {"path", 'p', "PATH", 0, "Path to config node", 2}},
//TODO		{'a' , {"data", 'a', "DATA", 0, "Data for create or update operation", 2}},
		{'c' , {"create", 'c', 0, 0, "create config node at PATH with DATA", 1}},
		{'r' , {"read", 'r', 0, 0, "read config node data from PATH", 1}},
//TODO		{'u' , {"update", 'u', 0, 0, "update config node at PATH as DATA", 1}},
		{'d' , {"delete", 'd', 0, 0, "delete config node at PATH", 1}},
```

a-petelin replied:
Что значит реально поддерживаемые?
Всё что надо это написать 3 строки кода про обновление

a-petelin replied:
Добавил выполнение update, завернул всё в try

dmerkushov commented on misc/ns-shutdown/main.cpp:
```diff
@@ -0,0 +1,47 @@
+
+int main(int argc, char** argv) {
+	map<char, NsCmdLineParameters::opt> prms = {
+		{'n' , {"name", 'n', "NAME", 0, "Name for shutdowning service"}},
```
```suggestion
		{'n' , {"name", 'n', "NAME", 0, "Name of service to shutdown"}},
```

dmerkushov commented on ns-skeleton/CMakeLists.txt:
```diff
@@ -63,14 +79,57 @@ set(Headers
+# Header-only MSGPACK realisation
+include_directories(${nanoservices_SOURCE_DIR}/third-party/msgpack-c/include)
+
+if(FILE)
+	include_directories(include/file)
+elseif(ZOO)
+	include_directories(include/zoo)
+	include_directories(/usr/include/zookeeper)
+	set(CONFIG_LIBS zookeeper_mt)
+else()
+	message(FATAL_ERROR "You cmake it wrong!!")
+endif(FILE)
```
Лучше директорию `file` не называть :) Чо-нть типа `config_file`, `config_zoo`...

a-petelin replied:
Переименовал

dmerkushov commented on ns-skeleton/CMakeLists.txt:
```diff
@@ -63,14 +79,57 @@ set(Headers
+	message(FATAL_ERROR "You cmake it wrong!!")
+endif(FILE)
+
+if(FILE)
+	add_definitions(-DURI="file:///etc/nanoservices/global.conf.json")
+	add_definitions(-DHEAD="/")	
+elseif(ZOO)
+	# Connection string to zookeeper instance
+	add_definitions(-DURI="zoo://127.0.0.1:2181")
+	add_definitions(-DHEAD="")
+	# Prefix for root config node in zookeeper
+	add_definitions(-DPREFIX="/ru/cniiag/webmaps")
+else()
+	message(FATAL_ERROR "You cmake it wrong!!")
+endif(FILE)
```
А зачем два ифа?

a-petelin replied:
Решил разбрать по блокам всё что зависит от флага. Иначе там бооольшой if с кучей всего будет.

dmerkushov commented on ns-skeleton/CMakeLists.txt:
```diff
@@ -103,23 +162,42 @@ if (DOXYGEN_FOUND)
+		${CMAKE_CURRENT_SOURCE_DIR}/test/AbstractConfigTest.h)
+	CXXTEST_ADD_TEST(config-tests config.cc ${Tests})
+	if(FILE)
+		include_directories(test/file)
```
Аналогично про имена директорий

dmerkushov commented on ns-skeleton/include/AbstractConfig.h:
```diff
@@ -0,0 +1,49 @@
+namespace nanoservices {
+	class AbstractConfig {
+	public:
+		/**
+		 *
+		 */
```
А что с комментариями случилось? :)

v-bus replied:
тут типа самодокументруемый код

a-petelin replied:
Добавил

dmerkushov commented on ns-skeleton/include/ConfigValidator.h:
```diff
@@ -0,0 +1,19 @@
+
+class ConfigValidator {
+public:
+	/**
+	 * 
+	 */
```
Коммент?

dmerkushov commented on ns-skeleton/include/CVInvalidType.h:
```diff
@@ -0,0 +1,8 @@
+#include "NsException.h"
+
+class CVInvalidType : public nanoservices::NsException {
```
Про эксепшены надо обсудить. Если уж их возвращать, то возвращать везде детальные.

И кстати, почему эти не в неймспейсе?

v-bus replied:
изначально это исключения конфигуратора, который внешний по отношению к скелетону.
у нас и не определено как инкапсулировать в самом скелетоне исключения. у нас уже 2+ сторонние проекта в коде скелетона, сторонние же не должны знать о скелетоновских исключениях.

dmerkushov replied:
Я не про неймспейс скелетона (какового неймспейса нет), я про неймспейс `nanoservices`, который для всего этого проекта

dmerkushov commented on ns-skeleton/include/NsCmdLineParameters.h:
```diff
@@ -37,10 +37,9 @@ namespace nanoservices {
 		 */
-		static std::shared_ptr<NsCmdLineParameters> instance() throw(nanoservices::NsException);
-
+		static std::shared_ptr<NsCmdLineParameters> instance () throw (nanoservices::NsException);
```
```suggestion
		static std::shared_ptr<NsCmdLineParameters> instance() throw (nanoservices::NsException);
```

dmerkushov commented on ns-skeleton/include/NsCmdLineParameters.h:
```diff
@@ -49,39 +48,48 @@ namespace nanoservices {
+		/**
+		 * Alias for type from getopt.h
+		 */
+		using opt = struct argp_option;
```
Есть уверенность, что именно `argp_option`?

a-petelin replied:
Там коммент старый

dmerkushov replied:
По-моему, стоит этот `uses` наружу вынести и использовать `opt` ещё и в `getOptionDefinitions()`

a-petelin replied:
В принципе вариант ...

dmerkushov commented on ns-skeleton/include/ConfigValidator.h:
```diff
@@ -0,0 +1,19 @@
+	 */
+	virtual void validate(const std::string& path, nanoservices::NsSkelJsonPtr data);
+	
+	static const int32_t childrenCount = 1024;
```
Лучше `MAX_CHILDREN_COUNT`

v-bus replied:
и MAX_DATA_SIZE

a-petelin replied:
Переименовал

dmerkushov commented on ns-skeleton/src/NsCmdLineParameters.cpp:
```diff
@@ -21,164 +21,149 @@
-			}
-
-			string key = fromNsSkelJsonPtr<string>(param->at("key"));
+char* new_c_str(std::string str) {
+	char* res = new char[str.size() + 1];
+	strcpy(res, str.c_str());
+	return res;
+}
```
М.б. эту функцию в NsUtils?

v-bus replied:
питоном повеяло)

a-petelin replied:
Скорее всего, как и извлекалку параметров коммандной строки куда нибудь в NsSkeleton

a-petelin replied:
Функцию перекинул

v-bus commented on misc/ns-shutdown/main.cpp:
```diff
@@ -0,0 +1,47 @@
+	map<char, NsCmdLineParameters::opt> prms = {
+		{'n' , {"name", 'n', "NAME", 0, "Name for shutdowning service"}},
+		{'p' , {"port", 'p', "PORT", 0, "Port of shutdowning service"}},
+		{'h' , {"host", 'h', "HOST", 0, "Host of shutdowning service"}},
```
```suggestion
		{'h' , {"host", 'h', "HOST", 0, "Host of service to shutdown"}},
```

v-bus commented on misc/ns-shutdown/main.cpp:
```diff
@@ -0,0 +1,47 @@
+int main(int argc, char** argv) {
+	map<char, NsCmdLineParameters::opt> prms = {
+		{'n' , {"name", 'n', "NAME", 0, "Name for shutdowning service"}},
+		{'p' , {"port", 'p', "PORT", 0, "Port of shutdowning service"}},
```
```suggestion
		{'p' , {"port", 'p', "PORT", 0, "Port of service to shutdown"}},
```

v-bus commented on misc/ns-shutdown/main.cpp:
```diff
@@ -0,0 +1,47 @@
+
+int main(int argc, char** argv) {
+	map<char, NsCmdLineParameters::opt> prms = {
+		{'n' , {"name", 'n', "NAME", 0, "Name for shutdowning service"}},
```
--all из NsSkelRpcRegistry
убить всех наносервисов

a-petelin replied:
Тогда он убьёт все общеизвестные в том числе и с других хостов, оно нам надо?

a-petelin replied:
Давайте на этот ключ отдельную issue  сделаем?

v-bus reviewed (commented):
third-party/cpp-httplib/fails/a.out а это кто?

a-petelin committed:
Fix english in misc/ns-shutdown/main.cpp

Co-Authored-By: Dmitriy Merkushov <dmerkushov@users.noreply.github.com>

a-petelin committed:
Fix english (2) in misc/ns-shutdown/main.cpp

Co-Authored-By: Viktor Bushmin <38643520+v-bus@users.noreply.github.com>

a-petelin committed:
Fix english (3) in misc/ns-shutdown/main.cpp

Co-Authored-By: Viktor Bushmin <38643520+v-bus@users.noreply.github.com>

a-petelin committed:
Refactor whitespace in ns-skeleton/include/NsCmdLineParameters.h

Co-Authored-By: Dmitriy Merkushov <dmerkushov@users.noreply.github.com>

a-petelin committed:
Fix english in README.md

Co-Authored-By: Dmitriy Merkushov <dmerkushov@users.noreply.github.com>

a-petelin committed:
Fix review suggesion about ns-configurator
 On branch ag-develop
	modified:   misc/ns-configurator/main.cpp

a-petelin committed:
Fix message in all CMakeLists.txt
 On branch ag-develop
	modified:   CMakeLists.txt
	modified:   misc/ns-configurator/CMakeLists.txt
	modified:   misc/ns-shutdown/main.cpp
	modified:   ns-skeleton/CMakeLists.txt

v-bus commented on README.md:
````diff
@@ -10,18 +10,37 @@ It is in development by now, a more meaningful readme will be added later
-make
+cmake --build .
 cd ../..
 ```
````
не все описано.
```
[ 55%] Building CXX object ns-skeleton/CMakeFiles/ns-skeleton.dir/src/zoo/ZooConfig.cpp.o
In file included from /home/v_bus/tmp/nanoservices/ns-skeleton/src/zoo/ZooConfig.cpp:9:
/home/v_bus/tmp/nanoservices/ns-skeleton/include/zoo/ZooConfig.h:2:10: fatal error: zookeeper.h: No such file or directory
 #include <zookeeper.h>
          ^~~~~~~~~~~~~
compilation terminated.
make[2]: *** [ns-skeleton/CMakeFiles/ns-skeleton.dir/build.make:375: ns-skeleton/CMakeFiles/ns-skeleton.dir/src/zoo/ZooConfig.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:91: ns-skeleton/CMakeFiles/ns-skeleton.dir/all] Error 2
make: *** [Makefile:152: all] Error 2
```

a-petelin replied:
Пофиксил описание зависимостей сборки

a-petelin committed:
Delete unneeded code snippets
 On branch ag-develop
	deleted:    third-party/cpp-httplib/fails/a.out
	deleted:    third-party/cpp-httplib/fails/main.cpp

a-petelin committed:
Fix README
 On branch ag-develop
	modified:   README.md
	new file:   misc/README.md

v-bus commented on misc/ns-shutdown/main.cpp:
```diff
@@ -0,0 +1,47 @@
+		{'n' , {"name", 'n', "NAME", 0, "Name of service to shutdown"}},
+		{'p' , {"port", 'p', "PORT", 0, "Port of service to shutdown"}},
+		{'h' , {"host", 'h', "HOST", 0, "Host of service to shutdown"}},
+		{'\0', {0,0,0,0}}
```
а это что значит \0?

a-petelin replied:
Я тут подумаю, может это будет само добавляться в конструкторе NsCmdLineParameters

a-petelin replied:
А в том классе написано зачем оно.

a-petelin replied:
Сделал

v-bus commented on ns-skeleton/include/NsUtils.h:
```diff
@@ -0,0 +1,14 @@
+	* @param ch -- delimeter character
+	*/
+	extern std::vector <std::string> string_split(std::string str, char ch);
+};
```
начался фреймворк...
а негде больше больше сплит стибрить?

a-petelin replied:
Неа ... я правда ещё в std::algoritms  не рылся

a-petelin replied:
Пока пусть так будет

v-bus commented on ns-skeleton/include/file/FileConfig.h:
```diff
@@ -0,0 +1,40 @@
+		FileConfig& operator=(FileConfig&) = delete;
+		void init(const std::string& configPath);
+
+		std::pair<nanoservices::NsSkelJsonPtr, std::string> goThroughPath(const std::string& path, nanoservices::NsSkelJsonPtr nodeobj);
```
вот тут мне без комментариев непонятно.

a-petelin replied:
Добавил

v-bus commented on ns-skeleton/include/zoo/ZooConfig.h:
```diff
@@ -0,0 +1,54 @@
+		
+		zhandle_t* _zooconnection;
+		std::shared_ptr<ConfigValidator> _validator;
+		static const char delimeter = '/';
```
тогда уж в дефайн, что ли. А то потеряется

a-petelin replied:
Думаю не стоит

v-bus commented on ns-skeleton/include/zoo/ZooConfig.h:
```diff
@@ -0,0 +1,54 @@
+		std::vector<std::string> _prefixpath;
+		void checkAndCreatePrefixPath();
+		
+		std::vector<char> typedByte = {'\0', '\0', '\0', 's', 'd', 'b'};
```
и сюда комментарий.

a-petelin replied:
Тоже

v-bus commented on ns-skeleton/src/NsSkelRpcHttpServer.cpp:
```diff
@@ -31,135 +31,56 @@
-
-void NsSkelRpcHttpServer::processIncomingConnection(int dataSocketFd) try {
+	_serverptr = make_shared<Server>();
+	_serverptr->Get(R"(/((?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$))", [&](const Request& req, Response& res) {
```
я не умею читать регексп сложнее .^.+
что там было написано?

a-petelin replied:
Это проверка на то что это валидный base64, т.е. если ткнуться чем-то что нельзя раскодировать в base64 то тебе прийдёт 404, а эта лямбда даже вызвана не будет

v-bus replied:
```suggestion
         /**
         * Check request contains base64
         */
	_serverptr->Get(R"(/((?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$))", [&](const Request& req, Response& res) {
```

dmerkushov replied:
Это коммент в коде, а не документационный. Поэтому надо не `/**`, а `//` (если бы был многострочный - `/*`)

a-petelin replied:
Я заделал свой вариант и там просто многострочный

v-bus replied:
да, этот мой)

v-bus commented on ns-skeleton/src/NsSkelRpcHttpServer.cpp:
```diff
@@ -31,135 +31,56 @@
-void NsSkelRpcHttpServer::processIncomingConnection(int dataSocketFd) try {
+	_serverptr = make_shared<Server>();
+	_serverptr->Get(R"(/((?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$))", [&](const Request& req, Response& res) {
+		shared_ptr<string> requestBase64 = make_shared<string>(req.matches[1]);
```
и сюда коммент

a-petelin replied:
+

v-bus commented on ns-skeleton/src/NsSkelRpcHttpServer.cpp:
```diff
@@ -31,135 +31,56 @@
-	sockstream >> httpVersion;
-
-	if (httpVersion.substr(0, 6) != "HTTP/1") {
+		// DEBUG
```
а может нормальное логгирование? у Шандра свой целый кастомный класс есть.

a-petelin replied:
Это остатки того что было до этого у Шандра, в принципе надо с ним обсудит

@dmerkushov и что нам с этим делать?

dmerkushov replied:
Лучше закомментировать. Это при отладке бывает нужно

v-bus commented on ns-skeleton/src/NsSkelRpcHttpServer.cpp:
```diff
@@ -31,135 +31,56 @@
-	cout << "processIncomingConnection(): Wait for response: " << waitForResponse << endl;
-	cout_lock.unlock();
+			res.set_content(*resultBase64, "application/base64");
+			res.set_header("Access-Control-Allow-Origin", "*");
```
а кто прислал? не будем проверять? а то может ходят тут всякие с рест тестилками, хедеров не знают

a-petelin replied:
Сейчас их в 90% случаев пошлёт, позже будет добавлены проверки не только на отправителя.

v-bus commented on ns-skeleton/src/NsSkelRpcHttpServer.cpp:
```diff
@@ -31,135 +31,56 @@
-		cout << "processIncomingConnection(): base64 result: " << *resultBase64 << endl;
-		cout_lock.unlock();
+void NsSkelRpcHttpServer::startup() throw(NsException) {
+	_serverThread = std::thread([&]() {this->_serverptr->listen("0.0.0.0", this->port());});
```
листен надо дефолт выставить и из конфига читать, нет?

a-petelin replied:
Теперь он читается из конфига

a-petelin committed:
Refactor move new_c_str() to NsUtils.cpp
	Add some doxy-comments
 On branch ag-develop
	modified:   ns-skeleton/include/NsCmdLineParameters.h
	modified:   ns-skeleton/include/NsUtils.h
	modified:   ns-skeleton/src/NsCmdLineParameters.cpp
	modified:   ns-skeleton/src/NsUtils.cpp

a-petelin committed:
Refactor rename folder from zoo/ and file to *_config
 On branch ag-develop
	modified:   misc/ns-configurator/CMakeLists.txt
	modified:   ns-skeleton/CMakeLists.txt
	renamed:    ns-skeleton/include/file/FileConfig.h -> ns-skeleton/include/file_config/FileConfig.h
	renamed:    ns-skeleton/include/zoo/ZooConfig.h -> ns-skeleton/include/zoo_config/ZooConfig.h
	renamed:    ns-skeleton/src/file/FileConfig.cpp -> ns-skeleton/src/file_config/FileConfig.cpp
	renamed:    ns-skeleton/src/zoo/ZooConfig.cpp -> ns-skeleton/src/zoo_config/ZooConfig.cpp
	renamed:    ns-skeleton/test/file/TestActionCheck.h -> ns-skeleton/test/file_config/TestActionCheck.h
	renamed:    ns-skeleton/test/zoo/TestActionCheck.h -> ns-skeleton/test/zoo_config/TestActionCheck.h

a-petelin committed:
Refactor fix old doxy-comment at opt alias
 On branch ag-develop
	modified:   ns-skeleton/include/NsCmdLineParameters.h

a-petelin committed:
Refactor doxy-comments now have mean
 On branch ag-develop
	modified:   ns-skeleton/include/AbstractConfig.h
	modified:   ns-skeleton/include/ConfigValidator.h
	modified:   ns-skeleton/src/ConfigValidator.cpp

a-petelin committed:
Refactor rename constants in ConfigValidator
 Your branch and 'origin/ag-develop' have diverged,
	modified:   ns-skeleton/include/ConfigValidator.h
	modified:   ns-skeleton/src/ConfigValidator.cpp

a-petelin committed:
Refactor move add zero terminator for argp option map to constructor
 On branch ag-develop
	modified:   misc/ns-configurator/main.cpp
	modified:   misc/ns-shutdown/main.cpp
	modified:   ns-skeleton/src/NsCmdLineParameters.cpp

a-petelin committed:
Fix constant listen at NsSkelRpcHttpServer, use host from config now
 On branch ag-develop
	modified:   ns-skeleton/include/NsSkelRpcServer.h
	modified:   ns-skeleton/src/NsSkelRpcHttpServer.cpp
	modified:   ns-skeleton/src/NsSkelRpcServer.cpp
	modified:   ns-tests/json.txt

a-petelin committed:
Refactor add some comments
 On branch ag-develop
	modified:   ns-skeleton/include/file_config/FileConfig.h
	modified:   ns-skeleton/include/zoo_config/ZooConfig.h
	modified:   ns-skeleton/src/NsSkelRpcHttpServer.cpp

dmerkushov commented on misc/README.md:
```diff
@@ -0,0 +1,9 @@
+
+## ns-configurator
+
+Tool for work with nanoservices config
```
Подробнее надо

a-petelin replied:
Это чтобы не пустой коммитить. Что писать я понятия не имею. Накидайте что вам не хватает знать и я распишу.

dmerkushov replied:
Чего тут не хватает... Как его юзать - не хватает. И ссылки из главного ридми сюда, чтобы при взгляде оттуда было видно, где про конфигурирование рассказано

a-petelin replied:
Ссылка из главного readme есть

dmerkushov commented on misc/ns-configurator/VERSION.txt:
```diff
@@ -0,0 +1 @@
+0.1
```
Давай версию ту же, что и у либы. Может, даже свой VERSION файл не нужен

a-petelin replied:
Этот файлик удалить забыл, я пробовал  покомпонентно назначать версии, но это не работает

a-petelin replied:
Удалено

dmerkushov commented on misc/testService/testMethod.h:
```diff
@@ -0,0 +1,24 @@
+
+class TestMethod: public nanoservices::NsSkelRpcReplier<TestMethodArgs, TestMethodResult> {
+public:
+	std::shared_ptr<TestMethodResult> processRequest(std::shared_ptr<TestMethodArgs> args) throw(nanoservices::NsException);
```
```suggestion
	std::shared_ptr<TestMethodResult> processRequest(std::shared_ptr<TestMethodArgs> args) noexcept(false);
```

Боюсь, придётся соответствовать общепринятой практике

a-petelin replied:
Ни вапрос, пройдусь везде и заменю.

Кстати auto_ptr тоже deprecated а в msgpack-c он есть. Возможно это изза директивы про Cxx03

a-petelin replied:
`noexcept(false)` в принципе указывать не надо, т.к. компилятору пофиг.

А вот `noexcept(true)` там где необходимо стоит указать

a-petelin replied:
Сделал

dmerkushov commented on ns-skeleton/FindNanoservices.cmake:
```diff
@@ -0,0 +1,62 @@
+endfunction(str_right)
+
+
+execute_process(COMMAND dpkg-query -W -f=\${Package}\\tProvides:\${Provides}\\n
+		COMMAND grep -E "nanoservices$"
+		COMMAND sed -e "s/\\\(.*\\\)\\t.*/\\1/"
+	OUTPUT_VARIABLE SKELETON_LIB_DEB)
```
```suggestion
execute_process(
	COMMAND dpkg-query -W -f=\${Package}\\tProvides:\${Provides}\\n
	COMMAND grep -E "nanoservices$"
	COMMAND sed -e "s/\\\(.*\\\)\\t.*/\\1/"
	OUTPUT_VARIABLE SKELETON_LIB_DEB
)
```

a-petelin replied:
Не понял что поменяно ...

dmerkushov replied:
> Не понял что поменяно ...

Скобочки-табуляции

dmerkushov commented on ns-skeleton/include/AbstractConfig.h:
```diff
@@ -0,0 +1,57 @@
+		virtual void update(const std::string& path, nanoservices::NsSkelJsonPtr data) = 0;
+		/**
+		 * Delete config node at path
+		 *   All subnodes delete too.
```
```suggestion
		 *   All subnodes will be deleted too.
```

dmerkushov commented on ns-skeleton/include/AbstractConfig.h:
```diff
@@ -0,0 +1,57 @@
+	class AbstractConfig {
+	public:
+		/**
+		 * Get instance of current realisation of AbstractConfig
```
```suggestion
		 * Get a singleton instance of the currently configured implementation of AbstractConfig
```
Синглтон же?

dmerkushov commented on ns-skeleton/include/AbstractConfig.h:
```diff
@@ -0,0 +1,57 @@
+		 */
+		static std::shared_ptr<AbstractConfig> instance();
+		/**
+		 * Initialisation connection to config source
```
```suggestion
		 * Connect to the configuration source
```

dmerkushov replied:
Может, и метод `connect()` назвать?

a-petelin replied:
Но с файлом то он connect не делает

dmerkushov replied:
Он коннектится к источнику конфигурации ^)

А что источник файловый, ну так он файловый

dmerkushov commented on ns-skeleton/include/AbstractConfig.h:
```diff
@@ -0,0 +1,57 @@
+		 */
+		virtual nanoservices::NsSkelJsonPtr read(const std::string& path, bool desc = false) = 0;
+		/**
+		 * Update config node at path use data
```
```suggestion
		 * Update config node at path using the given data
```

dmerkushov commented on ns-skeleton/include/AbstractConfig.h:
```diff
@@ -0,0 +1,57 @@
+		static void close();
+		/**
+		 * Create config node at path with data.
+		 *   If data define subnodes, they created too
```
```suggestion
		 *   If data has subnodes, they are created too
```

dmerkushov commented on ns-skeleton/include/ConfigValidator.h:
```diff
@@ -0,0 +1,26 @@
+class ConfigValidator {
+public:
+	/**
+	 * Valide data, path used only for generate errors messages
+	 *   Now check:
+	 *    - count of subnodes at one node
+	 *    - size of node data 
+	 *    - using appropriate types for define node data
+	 *    - definition of data for leaf subnodes
+	 * @param path
+	 * @param data
```
```suggestion
	 * Valide configuration data
	 *   By now, checks are:
	 * <ul>
	 * <li>count of subnodes at one node
	 * <li>size of node data 
	 * <li>usage of appropriate types to define node data
	 * <li>definition of data for leaf subnodes
	 * </ul>
	 * @param path used only to generate error messages
	 * @param data
```

dmerkushov commented on ns-skeleton/include/NsSkelRpcServer.h:
```diff
@@ -84,30 +84,34 @@ namespace nanoservices {
 
 		virtual void processIncomingConnection(int dataSocketFd);
 
+		std::thread _serverThread;
+		std::atomic<bool> _serverActive;
```
То есть я понимаю, что это сюда вытащено для того, чтобы дать детям доступ к этим полям. Но имхо если им нужен доступ, лучше дать им геттеры-сеттеры.

a-petelin replied:
Можно вернуть

a-petelin replied:
Вроде решили что закрываем

dmerkushov commented on ns-skeleton/include/NsSkelRpcService.h:
```diff
@@ -38,7 +38,7 @@ namespace nanoservices {
 						 bool httpEnabled, uint16_t httpPort);
 
-		NsSkelRpcService(NsSkelJsonPtr serviceJson) throw(NsException);
+		NsSkelRpcService(std::shared_ptr<std::string> serviceName, NsSkelJsonPtr serviceJson) throw(NsException);
```
Не уверен, что сюда стоит сервис нейм пихать, а не читать его из джсона

a-petelin replied:
Сервис нейма теперь нет в этом json

dmerkushov replied:
А почему теперь нету?

v-bus replied:
Мыж хотели какие-то другие вариаты lookup service делать. Я помню только те варианты, что мы с Лехой обсуждали, но там json в качестве диспатчера не было.

a-petelin replied:
В этом json нет имени потому что этот json теперь извлекается из словаря по имени сервиса а не из массива

dmerkushov commented on ns-skeleton/src/ConfigValidator.cpp:
```diff
@@ -0,0 +1,65 @@
+
+void ConfigValidator::validateObject(const std::string& path, const std::string& name, nanoservices::NsSkelJsonPtr data) {
+	switch(data->type()) {
+	case JSON_OBJECT: {
```
Сюда бы ещё `JSON_ARRAY`, у которого тоже дети есть, и про который вообще в этом свитче забыли :)

a-petelin replied:
`JSON_ARRAY` не поддерживается пока

dmerkushov replied:
Всё равно бы его сюда добавить закомменченный с тудушкой и пояснением

v-bus replied:
это @dmerkushov философский вопрос, как совать в ветку массивы. надо придумать кодировку
мы думали, может у тебя будут идеи

dmerkushov replied:
А чем не подходит:
"0" -> нулевой элемент
"1" -> первый
...

v-bus replied:
Вот и добавь свой вариант. Мы пока без JSON_ARRAY обходимся, строк хватает. Если ты считаешь, что он щас нужен - добавь с логикой.

v-bus replied:
мы сделали ограничения, ты видел, что детей не более 1024, в каждом листике не больше 1 Мб байтовый данных. 1Мб - рекомендация зукипера, ему так проще синхронизоваться. Поэтому просто засунуть массив в листик, значит, проверять в валидаторе, что он меньше 1Мб., а если резать ветками, то массив станет не больше 1024 элементов. Хотя это конфиги, и хрен знает, для чего больше массив использовать.

dmerkushov replied:
Я не знаю, зачем он может понадобиться. Если/когда понадобится - можно будет сделать так, как понадобится. Но когда я читал исходники, у меня возникло ощущение, что про него просто забыли. На данный момент достаточно было бы написать комментарий, что массив не поддерживаем, и в сообщении об ошибке об этом тоже честно говорить.

v-bus replied:
@a-petelin добавь JSON_ARRAY 1024 max element и 1Mb each element. И пусть портиться. Вдруг мы тут timeline начнем хранить эталонный.

a-petelin replied:
Проще добавить в других местах выдачу исключения если `JSON_ARRAY` встретился

a-petelin replied:
Добавил
https://github.com/dmerkushov/nanoservices/pull/18/commits/f20c33bf1f255471e28ffad7a596421a5d764111

dmerkushov commented on ns-skeleton/src/ConfigValidator.cpp:
```diff
@@ -0,0 +1,65 @@
+	}
+	case JSON_NUMBER:
+	case JSON_BOOLEAN:
+	case JSON_NULL:
```
А чо раньше скобочки есть, а тут нету? :)

a-petelin replied:
Переменные не объявляются

dmerkushov commented on ns-skeleton/src/NsSkelConfiguration.cpp:
```diff
@@ -30,8 +30,9 @@
-#include "NsSkelJson.h"
 #include "NsException.h"
+#include "AbstractConfig.h"
+
```
Строчка лишняя)

dmerkushov commented on ns-skeleton/src/file_config/FileConfig.cpp:
```diff
@@ -0,0 +1,182 @@
+
+NsSkelJsonPtr FileConfig::readWhole(const std::string& path) {
+	if(path[0] != '/') {
+		throw NsException(NSE_POSITION, "Path must start from \'/\'!");
```
```suggestion
		throw NsException(NSE_POSITION, "Path must start with \'/\'!");
```

dmerkushov commented on ns-skeleton/src/zoo_config/ZooConfig.cpp:
```diff
@@ -0,0 +1,303 @@
+	int port = 0;
+    
+	if (regex_match(uri, pieces_match, pieces_regex)) {
+		// Extract data from 1st parentsizes
+		host = pieces_match[1].str();
+		// Extract data from 2nd parentsizes
```
```suggestion
		// Extract data from 1st parentheses
		host = pieces_match[1].str();
		// Extract data from 2nd parentheses
```

dmerkushov commented on ns-skeleton/src/zoo_config/ZooConfig.cpp:
```diff
@@ -0,0 +1,303 @@
+	string url = host;
+	url += ":" + to_string(port);
+	// TODO: Use logger ??
+	auto log_fp_ = fopen("/dev/null", "w");
+	zoo_set_log_stream(log_fp_);
```
Мдяя...

a-petelin replied:
Это чистая заглушка. Надо принять решение что делать с тем логом что генерит либа общения с zookeeper.

dmerkushov replied:
Ну, прям как хорошо бы было - это гнать лог зукипера в какой-то поток в памяти, а потом этот поток засовывать в наш логгер по мере поднятия соединения с ним

a-petelin replied:
Для этого надо придумать как получить файловый дескриптор на этот поток ....

v-bus replied:
@a-petelin так это настройка zookeeper самого или zoo.h? 
@dmerkushov  я так понял, что это настройка для локально установленного зукипера и кроме дебага не имеет смысла
и мне не нравится закоменченный дебаг строки. это срр у вас тут все равно все в макросах

dmerkushov replied:
@a-petelin 
> Для этого надо придумать как получить файловый дескриптор на этот поток
* файловая система в памяти
* `/tmp`
* named pipe

@v-bus что такое "закомменченный дебаг строки"? :)

v-bus replied:
@a-petelin ну Шандр предлагал оставить кусок отработки RPC //DEBUG в комментариях

a-petelin replied:
@v-bus этот лог это выхлоп библиотеки что ходит к `zookeeper`

a-petelin replied:
В итоге решили завести отдельный issue, там обсудить и делать как решили

dmerkushov commented on ns-skeleton/src/zoo_config/ZooConfig.cpp:
```diff
@@ -0,0 +1,303 @@
+	thread([&](){
+		this_thread::sleep_for(std::chrono::milliseconds(timeoutms));
+		if(state == -1) {
+			connect_wait.set_exception(make_exception_ptr(NsException(NSE_POSITION, "Coudn't connect to zookeeper!")));
```
```suggestion
			connect_wait.set_exception(make_exception_ptr(NsException(NSE_POSITION, "Couldn't connect to zookeeper!")));
```

dmerkushov commented on ns-skeleton/src/zoo_config/ZooConfig.cpp:
```diff
@@ -0,0 +1,303 @@
+	zoo_set_log_stream(log_fp_);
+	zoo_set_debug_level(ZOO_LOG_LEVEL_DEBUG);
+	promise<int> connect_wait;
+	// TODO: move to constant
+	int timeoutms = 3000;
```
Я даже больше скажу: не на константу надо заменить, а на параметр конфигурации

a-petelin replied:
Не уверен. Это чисто магия, для того чтобы не долбиться полчаса в пингуемый хост где zookeeper отсутствует.

dmerkushov replied:
Я понимаю, зачем эта магия, но таймауты нужно оптимизировать не только для того чтобы не долбиться туда, где сервис мёртв, но и чтобы вовремя обнаруживать ошибки. Этим делом админы занимаются. Если зукипер в какой-то инсталляции либо отвечает за 100 мс, либо не отвечает никогда, то ставить таймаут 3000 мс может быть расточительно, пусть лучше наши сервисы умирают поскорее: раньше на приборах станет видно. Так что это должно конфигуриться, по-моему. Как дефолт 3000 норм, разумеется

a-petelin replied:
Тут всё сложнее чем ты посчитал.

v-bus replied:
управление таймаутами это отдельный геммор. Давайте в конфиг.

a-petelin replied:
Конкретно этим таймаутом я бы не управлял. Тут словами рассказывать надо, а не писать буквами больше чем я уже кода написал.

a-petelin replied:
В итоге решили завести отдельный issue, там обсудить и делать как решили

dmerkushov commented on ns-skeleton/src/zoo_config/ZooConfig.cpp:
```diff
@@ -0,0 +1,303 @@
+void ZooConfig::checkAndCreatePrefixPath() {
+	string path = "";
+	for(auto p: _prefixpath) {
+		path += delimeter + p;
```
`delimiter` через **i**

a-petelin replied:
Готово

dmerkushov commented on ns-skeleton/src/zoo_config/ZooConfig.cpp:
```diff
@@ -0,0 +1,303 @@
+
+void ZooConfig::createDict(const std::string& path, const std::string& parent_path, const std::string name, nanoservices::NsSkelJsonPtr data) {
+	switch(data->type()) {
+	case JSON_OBJECT: {
```
Опять про `JSON_ARRAY` забыли )

a-petelin replied:
См. выше

a-petelin replied:
Добавил
https://github.com/dmerkushov/nanoservices/pull/18/commits/f20c33bf1f255471e28ffad7a596421a5d764111

dmerkushov commented on ns-skeleton/src/zoo_config/ZooConfig.cpp:
```diff
@@ -0,0 +1,303 @@
+}
+
+vector<string> ZooConfig::getChildren(const string& path) {
+	promise<vector<string>> child_v_p;
+	auto child_v_f = child_v_p.get_future();
```
`v_p`, `v_f` ... Попонятнее бы

a-petelin replied:
Promise и его future

dmerkushov commented on ns-skeleton/src/zoo_config/ZooConfig.cpp:
```diff
@@ -0,0 +1,303 @@
+		break;
+	default:
+		stringstream ess;
+		ess <<  "Data at \"" << path << "\" corrupted or used unknown type specifier!";
```
```suggestion
		ess <<  "Data at \"" << path << "\" is corrupt or uses an unknown type specifier!";
```

dmerkushov commented on ns-skeleton/src/zoo_config/ZooConfig.cpp:
```diff
@@ -0,0 +1,303 @@
+
+void ZooConfig::updateDict(const std::string& path, nanoservices::NsSkelJsonPtr data) {
+	switch(data->type()) {
+	case JSON_OBJECT: {
```
Ну а как же `JSON_ARRAY`?

a-petelin replied:
См. выше

a-petelin replied:
Добавил
https://github.com/dmerkushov/nanoservices/pull/18/commits/f20c33bf1f255471e28ffad7a596421a5d764111

a-petelin committed:
Fix english part 4

Co-Authored-By: Dmitriy Merkushov <dmerkushov@users.noreply.github.com>

dmerkushov commented on README.md:
```diff
@@ -10,18 +10,38 @@ It is in development by now, a more meaningful readme will be added later
 Header files to include to use the library will be in `nanoservices/ns-skeleton/include/`
+
+A successful build will also generate some tools in `nanoservices/build/misc/` folder. 
+More info about [here](/misc/README.md)
```
Напиши плз конкретно, что там есть тулза для конфигурирования и для старта-останова. А подробности читайте там-то

v-bus replied:
Дорогая редакция, сам нажми кнопку insert a suggestion и напиши. все равно потом будешь редактировать, что два раза мучаться.

a-petelin replied:
И так каждую тулзу сюда прописывать?

dmerkushov replied:
```suggestion
By now, these are tools for configuring nanoservices (ns-configurator) and shutdowning a started nanoservice remotely (ns-shutdown). [More info](/misc/README.md)

It is highly recommended to use the supplied tools to configure nanoservices, as different configuration source strategies may be applied by the nanoservice hosting teams.
```

a-petelin committed:
Delete unneeded version file
 On branch ag-develop
	deleted:    misc/ns-configurator/VERSION.txt

a-petelin committed:
Merge branch 'ag-develop' of https://github.com/dmerkushov/nanoservices into ag-develop

a-petelin committed:
Refactor remove throw()
	fix lib package
	fix testService
 On branch ag-develop
	modified:   misc/testService/CMakeLists.txt
	modified:   misc/testService/testMethod.cpp
	modified:   misc/testService/testMethod.h
	modified:   ns-skeleton/CMakeLists.txt
	modified:   ns-skeleton/FindNanoservices.cmake
	new file:   ns-skeleton/postinst

dmerkushov committed:
+ README on ns-configurator

a-petelin committed:
Refactor add tools info into README.md

Co-Authored-By: Dmitriy Merkushov <dmerkushov@users.noreply.github.com>

a-petelin committed:
Refactor add exception for unused JSON types
 On branch ag-develop
	modified:   ns-skeleton/src/zoo_config/ZooConfig.cpp

a-petelin committed:
Merge branch 'ag-develop' of https://github.com/dmerkushov/nanoservices into ag-develop

a-petelin committed:
Refartor fix english in variable names
 On branch ag-develop
	modified:   ns-skeleton/include/zoo_config/ZooConfig.h
	modified:   ns-skeleton/src/zoo_config/ZooConfig.cpp

a-petelin committed:
Remove c++17 feature

a-petelin committed:
Up version

a-petelin committed:
Merge pull request #21 from dmerkushov/19-bug-build-failed-v6.2.0

19 bug build failed v6.2.0

v-bus committed:
Added Dockerfile image builds nanoservice from source
Added Docker Compose v3 YAML
Image on Ubuntu 18.04 with all dependecies to build project
On branch v-bus-add-docker-ubuntu18-image-and-container
new file:   supplemental/docker/ubuntu18build/Dockerfile
new file:   supplemental/docker/ubuntu18build/README.md
new file:   supplemental/docker/ubuntu18build/docker-compose.yml

v-bus committed:
Update README.md

dmerkushov committed:
Merge pull request #22 from dmerkushov/v-bus-add-docker-ubuntu18-image-and-container

Added Dockerfile image builds nanoservice from source
